sn_cli/src/files/files_uploader.rs

```diff
@@ -0,0 +1,337 @@
+        let entries = WalkDir::new(path).into_iter().flatten();
+        self.entries_to_upload.extend(entries);
+        self
+    }
+
+    pub async fn start_upload(self) -> Result<UploadSummary> {
+        let mut chunk_manager = ChunkManager::new(&self.root_dir);
+        let chunks_to_upload = self.get_chunks_to_upload(&mut chunk_manager).await?;
+        let chunks_to_upload_len = chunks_to_upload.len();
+
+        let now = Instant::now();
+
+        let mut uploader = Uploader::new(self.client, self.root_dir);
+        uploader.set_upload_cfg(self.upload_cfg.clone());
+        uploader.insert_chunk_paths(chunks_to_upload);
+
+        let events_handle = Self::spawn_upload_events_handler(
+            chunk_manager,
+            self.make_data_public,
+            chunks_to_upload_len,
+            uploader.get_event_receiver(),
+        )?;
+
+        for path in self.file_paths_to_print.iter() {
+            debug!(
+                "Uploading file(s) from {path:?} batch size {:?} will verify?: {}",
+                self.upload_cfg.batch_size, self.upload_cfg.verify_store
+            );
+            if self.make_data_public {
+                info!("{path:?} will be made public and linkable");
+                println!("{path:?} will be made public and linkable");
+            }
+        }
+        if self.file_paths_to_print.len() == 1 {
+            println!(
+                "Splitting and uploading {:?} into {chunks_to_upload_len} chunks",
+                self.file_paths_to_print[0]
+            );
+        } else {
+            println!(
+                "Splitting and uploading {:?} into {chunks_to_upload_len} chunks",
+                self.file_paths_to_print
+            );
+        }
+
+        let summary = match uploader.start_upload().await {
+            Ok(summary) => summary,
+            Err(ClientError::Wallet(WalletError::Transfer(TransferError::NotEnoughBalance(
+                available,
+                required,
+            )))) => {
+                return Err(eyre!(
+                    "Not enough balance in wallet to pay for chunk. \
+            We have {available:?} but need {required:?} to pay for the chunk"
+                ))
+            }
+            Err(err) => return Err(eyre!("Failed to upload chunk batch: {err}")),
+        };
+        events_handle.await??;
+
+        let elapsed = Self::msg_format_elapsed_time(now.elapsed());
+        println!(
+            "Among {chunks_to_upload_len} chunks, found {} already existed in network, uploaded \
+            the leftover {} chunks in {elapsed}",
+            summary.skipped_count, summary.uploaded_count,
+        );
+        info!(
+            "Among {chunks_to_upload_len} chunks, found {} already existed in network, uploaded \
+            the leftover {} chunks in {elapsed}",
+            summary.skipped_count, summary.uploaded_count,
+        );
+        println!("**************************************");
```

Comment:
This is message code. We need to extract this.

Reply:
Hey Jason, the structs from the CLI is now exposed as a `library` and hence I feel like we should not be ideally printing things to stdout. 
We can probably make these into events / have an interface that can write to std or probably do something else.

Also, I'm not sure if separating those into their own functions help when modifying the code. I felt they were very hard to track.

Reply:
Also, since the code is a short one, are so many redirections required?

Reply:
Message code is a cross cutting concern and we're not dealing with it correctly. It's going to pad up the codebase and I don't think there is a good justification for this.

Reply:
I feel like we can probably have events that callers of `FilesUploder` can consume to get these updates? And they would be responsible for printing the message (cli) or updating some progress bar (gui)?

Reply:
Do what you think is best, but be aware that message code like this is a [cross cutting concern](https://www.linkedin.com/advice/0/how-can-design-patterns-solve-cross-cutting-concerns#:~:text=Design%20patterns%20can%20help%20you%20solve%20cross%2Dcutting%20concerns%20by,key%20principle%20of%20software%20design.)

In Java. Easy. Rust - Not so sure. I will look into this.